syntax = "proto3";

package serverComm;

service SimulationDealer {
    rpc RequestTickUpdates(RobotRequest) returns (stream RobotData) {}
    rpc SendWriteInfo(RobotWrite) returns (WriteResult) {}
    rpc RequestServer(ServerRequest) returns (ServerResult) {}
    rpc RequestConnect(ClientRequest) returns (ClientResult) {}
    rpc RequestSend(SendRequest) returns (SendResult) {}
    rpc RequestRecv(RecvRequest) returns (RecvResult) {}
    rpc RequestGetClient(GetClientRequest) returns (GetClientResult) {}
}

message RobotRequest {
    string robot_id = 1;
}

message RobotData {
    int32 tick = 1;
    int32 tick_rate = 2;
    string content = 3;
}

message RobotWrite {
    string robot_id = 1;
    string attribute_path = 2;
    string value = 3;
}

message WriteResult {
    bool result = 1;
}

// Communications Messages

message ServerRequest {
    string robot_id = 1;
    string address = 2;
    string port = 3;
}

message ServerResult {
    bool result = 1;
    string msg = 2;
}

message ClientRequest {
    string robot_id = 1;
    string address = 2;
    string port = 3;
}

message ClientResult {
    string host_robot_id = 1;
    bool result = 2;
    string msg = 3;
}

message SendRequest {
    string robot_id = 1;
    string address = 2;
    string port = 3;
    string data = 4;
    string client_id = 5;
}

message SendResult {
    bool result = 1;
    string msg = 2;
}

message RecvRequest {
    string robot_id = 1;
    string address = 2;
    string port = 3;
    string client_id = 4;
}

message RecvResult {
    string data = 1;
    bool result = 2;
    string msg = 3;
}

message GetClientRequest {
        string robot_id = 1;
        string address = 2;
        string port = 3;
}

message GetClientResult {
    string client_id = 1;
    bool result = 2;
    string msg = 3;
}
